Nruof/gentools port

EtherealMC-Bit opened the pull request:
This PR ports the functionality of being able to calculate dt using the function from gentools from Warp. If DT is not passed to diode setup, it will use this function to calculate it.

nruof committed:
gentools from warp

nruof committed:
Remove functions not used in warpx

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/gentools_port

nruof committed:
Remove some warp declarations

EtherealMC-Bit committed:
Porting gentools function

- Removed a lot of uneeded functionality from gentools
- Removed the class containing a lot of functionality
  and moved the function we want out of it.

EtherealMC-Bit committed:
Moved function into util.py and delete gentools

EtherealMC-Bit committed:
Added vmax definition

EtherealMC-Bit committed:
remove unneeded imports

EtherealMC-Bit committed:
Added remaining functionality and test for calculating dt

EtherealMC-Bit committed:
Added to docstring and changed order of input vars

KZhu-ME commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -39,6 +39,10 @@ class DiodeRun_V1(object):
     # ### VACUUM ###
     # Cathode anode distance
     D_CA = 1e-3
+    # Vacuum bias of the highest voltage grid relative to the cathode
+    V_GRID = 100
+    # If calculating DT using a CFL factor, this is the CFL factor
+    CFL_FACTOR = 1
```
Should these values be none, since we're not supposed to have default values for these?

EtherealMC-Bit replied:
Yeah this would be better as `None` because there's no use for them unless you're doing the dt calculation

EtherealMC-Bit committed:
Edited docstring description

KZhu-ME commented on mewarpx/mewarpx/utils_store/parallel_util.py:
```diff
@@ -22,6 +18,7 @@ def mpiallreduce(data=None, op=None, comm=None):
     return result
 
 def parallelsum(a, comm=None):
+    from mewarpx.mwxrun import mwxrun
```
The reason mwxrun was imported for the whole file rather than this function is that there are other functions in warp's parallel_util that check if there is more than 1 processor `mwxrun.n_procs <= 1`.

EtherealMC-Bit replied:
So I moved it because if it's imported for the whole file the simulation crashes. Figuring that out is what delayed this PR a little.

KZhu-ME replied:
Oh wow that's weird...

KZhu-ME commented on mewarpx/mewarpx/utils_store/util.py:
```diff
@@ -387,3 +388,35 @@ def plasma_Debye_length(T, n):
+            dt_local = min(dx, dz) / vmax * CFL_factor
+        else:
+            dt_local = dz / vmax * CFL_factor
+        return parallel_util.mpiallreduce(data=dt_local, op=MPI.MIN)
```
To save on having to import MPI for util, we could change `mpiallreduce` so that it takes in a string as an argument for `op`. For example 
```

def mpiallreduce(data=None, opstring=None, comm=None):
    if opstring is None:
        op = mpi.SUM
    elif opstring == "SUM":
        op = mpi.SUM:
    elif opstring == "MIN":
        op = MPI.MIN
    else:
        raise NotImplementedError("The opstring is unrecognized or has not been implemented yet.")
    if comm is None:
        comm = comm_world
    # --- "fast" version was removed because it produced bugs
    if isinstance(data, np.ndarray) and data.dtype is not np.dtype('object'):
        result = np.empty_like(data)
        comm.Allreduce(data, result, op=op)
    else:
result = comm.allreduce(data, op=op)
```

EtherealMC-Bit replied:
Is there a negative performance impact or some other reason we want to avoid importing MPI?

KZhu-ME replied:
Overall I don't think it wouldn't be that big of an impact since the import is just being called once. But this does mean that runs that don't use mpi4py, would still have mpi4py cached as it is imported in this file.

EtherealMC-Bit replied:
Any thoughts on this change @roelof-groenewald ?

EtherealMC-Bit replied:
> Overall I don't think it wouldn't be that big of an impact since the import is just being called once. But this does mean that runs that don't use mpi4py, would still have mpi4py cached as it is imported in this file.

Yeah that's a good point

roelof-groenewald replied:
I think not needing to import mpi4py for non-parallel runs would be good so I'd say lets go with Kevin's suggestion.

KZhu-ME reviewed (changes requested):
Looks good! Just a few suggestions.

EtherealMC-Bit committed:
A couple of suggested changes

roelof-groenewald commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -39,6 +39,10 @@ class DiodeRun_V1(object):
     # ### VACUUM ###
     # Cathode anode distance
     D_CA = 1e-3
+    # Vacuum bias of the highest voltage grid relative to the cathode
+    V_GRID = None
```
This can be removed after implementing the changes discussed below. Main thing is that a diode cannot have a grid - then it would at minimum be a triode.

roelof-groenewald commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -340,6 +340,21 @@ def init_base(self):
             '''
 
+        if self.DT is None:
+            if self.V_GRID is None or self.CFL_FACTOR is None:
+                raise ValueError("Both V_GRID and CFL_FACTOR must be defined in order to calculate dt")
```
This should be changed to only check if `CFL_FACTOR` is specified.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/util.py:
```diff
@@ -387,3 +388,35 @@ def plasma_Debye_length(T, n):
         constants.epsilon_0 * constants.kb_J * T / (n * constants.e**2)
     )
+
+def calc_timesteps(V_anode, V_grid, CFL_factor, nx, nz, xmin, xmax, zmin, zmax):
```
We should make it that `V_grid` is an optional parameter (just use default `None`) since not all simulations will have an auxiliary grid. If `V_grid` is not specified it's value can be set to `5` (this will give a safe value for `vmax` in the case that the electron velocities are only set by their temperature).

roelof-groenewald reviewed (changes requested):
Some quick changes but mostly looks good.

EtherealMC-Bit committed:
Suggested changes from code review

roelof-groenewald commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -340,6 +338,20 @@ def init_base(self):
 
+        if self.DT is None:
+            if self.CFL_FACTOR is None:
+                raise ValueError("Both V_GRID and CFL_FACTOR must be defined in order to calculate dt")
```
```suggestion
                raise ValueError("CFL_FACTOR must be defined in order to calculate dt")
```

EtherealMC-Bit committed:
Fix error text for CFL factor not being defined.

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

EtherealMC-Bit committed:
Changed the 'op' parameter in parallel_util.mpiallreduce()

EtherealMC-Bit committed:
Merge branch 'nruof/gentools_port' of github.com:ModernElectron/WarpX into nruof/gentools_port

roelof-groenewald commented on mewarpx/mewarpx/utils_store/parallel_util.py:
```diff
@@ -1,15 +1,18 @@
 
-def mpiallreduce(data=None, op=None, comm=None):
-    if op is None:
+def mpiallreduce(data=None, opstring=None, comm=None):
+    if opstring is None:
```
Is there a reason to not just make the default value for `opstring="SUM"` and remove the `if opstring is None` block?

roelof-groenewald commented on mewarpx/mewarpx/utils_store/util.py:
```diff
@@ -387,3 +386,37 @@ def plasma_Debye_length(T, n):
         constants.epsilon_0 * constants.kb_J * T / (n * constants.e**2)
     )
+
+def calc_timesteps(V_anode, CFL_factor, nx, nz, xmin, xmax, zmin, zmax, V_grid=5):
+        """Calculate timestep size based on grid data and CFL parameter
```
Seeing how this has shaken out I think it is better to move this function to `mwxrun.py` and have it be a time counterpart to the `init_grid()` function. The reason is that it is unfortunate how would now have to call `mwxrun.init_grid(`xmin`, `xmax`, `zmin`, `zmax`, `nx`, `nz`)` with all the appropriate grid parameters and then again call `util.calc_timesteps()` with basically all the same parameters again. Instead we can have a function in `mwxrun.py` called `init_timestep` that takes either a fixed `DT` of a `CFL_factor` as input and if the `CFL_factor` is given it will calculate `DT` using the already defined grid quantities in `mwxrun`. Does that make sense?

EtherealMC-Bit committed:
Refactoring from code review

- Changed the default value of 'opstring' for mpiallreduce to be "SUM" instead of None.
- Moved the functionality for calculating dt into mwxrun.

EtherealMC-Bit committed:
Removed duplicate 'setting up simulation' printout

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -56,6 +58,51 @@ def init_grid(self, xmin, xmax, zmin, zmax, nx, nz, max_grid_size=4):
+               2*np.e/constants.m_e * max(0.0, V_anode, V_grid)
+        )
+
+        if self.grid.nx > 0:
+            dx = (self.grid.xmax - self.grid.xmin) / self.grid.nx
+        else:
+            dx = 0.
+        # The picmi standard uses x and y for 2d, while we use x and z
+        dz = (self.grid.ymax - self.grid.ymin) / self.grid.ny
+
+        if dx > 0.:
+            dt_local = min(dx, dz) / vmax * CFL_factor
+        else:
+            dt_local = dz / vmax * CFL_factor
```
Since this function has to be run after `init_grid()` (which also runs `_set_grid_params()`) you can just use `self.dx` instead of recalculating dx, and same for dz. Also, WarpX doesn't have functionality for 1D runs so you don't need the `if dx <= 0` part of the call below. (I didn't test this suggestion so please confirm that it works)
```suggestion
        if self.geom_str == 'XZ' or self.geom_str == 'RZ' :
            dt_local = min(self.dx, self.dz) / vmax * CFL_factor
        if self.geom_str == 'XYZ':
            dt_local = min(self.dx, self.dy, self.dz) / vmax * CFL_factor
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -56,6 +58,51 @@ def init_grid(self, xmin, xmax, zmin, zmax, nx, nz, max_grid_size=4):
+                will be calculated.
+            CFL_factor (float): Multiplier to determine the actual timestep
+                given the CFL ratio. eg. ``dt = CFL_factor * CFL_ratio``.
+            V_grid (float): Vaccum bias of highest-voltage grid relative to
+                cathode, V. If not defined a value of 5 which is a safe value
+                for finding vmax in the case that the electron velocities are
+                only set by their temperature
```
There seems to be a typo in the description here.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/util.py:
```diff
@@ -9,12 +9,11 @@
 from mewarpx.utils_store import mwxconstants as constants
 
 
+
```
```suggestion
```

EtherealMC-Bit committed:
Suggested changes

roelof-groenewald reviewed (approved):
Looks good. Nice work!

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -56,6 +58,45 @@ def init_grid(self, xmin, xmax, zmin, zmax, nx, nz, max_grid_size=4):
+            raise ValueError("No CFL_factor passed into init_timestep when DT is not defined!")
+
+        vmax = np.sqrt(
+               2*np.e/constants.m_e * max(0.0, V_anode, V_grid)
```
Oh, I missed one thing here. The voltages can be positive or negative so we should allow for that.
```suggestion
               2*np.e/constants.m_e * max(0.0, abs(V_anode), abs(V_grid))
```

KZhu-ME commented on mewarpx/mewarpx/utils_store/parallel_util.py:
```diff
@@ -1,15 +1,16 @@
-def mpiallreduce(data=None, op=None, comm=None):
-    if op is None:
+def mpiallreduce(data=None, opstring="SUM", comm=None):
+    if opstring is None:
```
```suggestion
    if opstring is None or opstring == "SUM":
```
I think with how it currently is, not passing an opstring, ie `opstring="SUM"` throws a NotImplementedError

KZhu-ME reviewed (changes requested):
Everything looks good, just one small issue. I think the pr can be merged once this is fixed.

EtherealMC-Bit committed:
Use absolute value for voltages and fixed issue in mpiallreduce
